lib/elixir/lib/process.ex

```diff
@@ -607,7 +609,7 @@ defmodule Process do
   Inlined by the compiler.
   """
   @spec hibernate(module, atom, list) :: no_return
-  def hibernate(mod, fun, args) do
+  def hibernate(mod, fun, args) when is_atom(mod) and is_atom(fun) and is_list(args) do
```

Comment:
I don't think we should add the guards here because both are inlined by the compiler. This means that:

    Process.hibernate(:foo, 1, [])

and

    x = Process
    x.hibernate(:foo, 1, [])

would have different stacktraces. We could use `defdelegate` if that makes developers less compelled to add guards.